drivers/dai/nxp/micfil/micfil.c

```diff
@@ -0,0 +1,202 @@
+struct micfil_bespoke_config {
+	uint32_t pdm_rate;
+	uint32_t pdm_ch;
+};
+
+static void dai_nxp_micfil_trigger_start(const struct device *dev)
+{
+	struct dai_nxp_micfil_data *micfil_data = dev->data;
+
+	PDM_Reset(micfil_data->base);
+	/* enable DMA requests */
+	PDM_EnableDMA(micfil_data->base, true);
+	/* enable the module */
+	PDM_Enable(micfil_data->base, true);
+}
+
+static void dai_nxp_micfil_trigger_stop(const struct device *dev)
+{
+	struct dai_nxp_micfil_data *micfil_data = dev->data;
+
+	/* disable DMA requests */
+	PDM_EnableDMA(micfil_data->base, false);
+	/* disable module */
+	PDM_Enable(micfil_data->base, false);
+}
+
+static const struct dai_properties
+*dai_nxp_micfil_get_properties(const struct device *dev, enum dai_dir dir, int stream_id)
+{
+	const struct dai_nxp_micfil_config *cfg = dev->config;
+
+	if (dir == DAI_DIR_RX) {
+		return cfg->rx_props;
+	}
+
+	LOG_ERR("%s: invalid direction %d", __func__, dir);
+	return NULL;
+}
+
+static int dai_nxp_micfil_trigger(const struct device *dev, enum dai_dir dir,
+				  enum dai_trigger_cmd cmd)
+{
+	if (dir != DAI_DIR_RX) {
+		LOG_ERR("%s: invalid direction %d", __func__, dir);
+		return -EINVAL;
+	}
+
+	switch (cmd) {
+	case DAI_TRIGGER_START:
+		dai_nxp_micfil_trigger_start(dev);
+		break;
+	case DAI_TRIGGER_STOP:
+	case DAI_TRIGGER_PAUSE:
+		dai_nxp_micfil_trigger_stop(dev);
+		break;
+	case DAI_TRIGGER_PRE_START:
+	case DAI_TRIGGER_COPY:
+		return 0;
+	default:
+		LOG_ERR("%s: invalid trigger cmd %d", __func__, cmd);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int dai_nxp_micfil_get_config(const struct device *dev, struct dai_config *cfg,
+				     enum dai_dir dir)
+{
+	struct dai_nxp_micfil_data *micfil_data = dev->data;
+
+	memcpy(cfg, &micfil_data->cfg, sizeof(*cfg));
+	return 0;
+}
+
+static int dai_nxp_micfil_set_config(const struct device *dev,
+		const struct dai_config *cfg, const void *bespoke_cfg)
+
+{
+	struct dai_nxp_micfil_data *micfil_data = dev->data;
+	const struct dai_nxp_micfil_config *micfil_cfg = dev->config;
+	const struct micfil_bespoke_config *bespoke = bespoke_cfg;
+	pdm_config_t global_config = { 0 };
+	pdm_channel_config_t chan_config = { 0 };
+	int ret;
+	int i;
+
+	if (cfg->type != DAI_IMX_MICFIL) {
+		LOG_ERR("wrong DAI type: %d", cfg->type);
+		return -EINVAL;
+	}
+
+	global_config.fifoWatermark = micfil_cfg->fifo_depth - 1;
+	global_config.qualityMode = kPDM_QualityModeVeryLow0;
+	global_config.cicOverSampleRate = MICFIL_OSR_DEFAULT;
+
+	PDM_Init(micfil_data->base, &global_config);
+
+	for (i = 0; i < bespoke->pdm_ch; i++) {
+		chan_config.gain = kPDM_DfOutputGain2;
+		chan_config.cutOffFreq = kPDM_DcRemoverBypass;
+		PDM_SetChannelConfig(micfil_data->base, i, &chan_config);
+	}
+	ret = PDM_SetSampleRateConfig(micfil_data->base, MICFIL_CLK_ROOT, bespoke->pdm_rate);
+	if (ret != kStatus_Success) {
+		LOG_ERR("Failure to set samplerate config rate %d", bespoke->pdm_rate);
+		return ret;
```

Comment:
doesn't this need conversion to errno-like code? is ret < 0?

Reply:
True, this function can only return kStatus_Success or KStatus_Fail. Will return `-EINVAL` in case of failure.